### ISSUE TITLE
A question about DefaultResourceLeak's constructor

### ISSUE THREAD

**Issue by carryxyh** [netty/netty#8462]
Hi, community
In the constructor of DefaultResourceLeak, I see this sentence: It's important that we not store a reference to the referent as this would disallow it from be collected via the WeakReference.

However, allLeaks.put(this, LeakEntry.INSTANCE); put 'this' as the key, which will cause the relationship of allLeaks -> DefaultResourceLeak to always exist, which does not match the description of doc.

Also, I did a simple test with the following code:
```
        ReferenceQueue<Object> r = new ReferenceQueue<Object>();

        ConcurrentMap<WeakReference<?>, WeakReference> c = PlatformDependent.newConcurrentHashMap();
        WeakReference<Object> w = new WeakReference<Object>(new Object(), r);

        // this two line will cause the same result.
        c.put(w, w);
        c.put(w, new WeakReference(new Object()));

        System.gc();
        Thread.sleep(1000);
        WeakReference<Object> will = (WeakReference<Object>) r.poll();
        // will print true means that 'w' will be collected by jvm.
        System.out.println(w == will);
```

Here 'c' is equivalent to allLeaks, I found that c.put(w, w); and c.put(w, new WeakReference(new Object())); do not affect w (where 'w' can be considered a DefaultResourceLeak instance) Recycling.

In summary, I think the comments in the DefaultResourceLeak constructor are incorrect, allLeaks.put(this, LeakEntry.INSTANCE); and allLeaks.put(this, this); there is actually no difference.

How do u think about it?

**Comment by carryxyh** [comment]
```
// this two line will cause the same result.
c.put(w, w);
c.put(w, new WeakReference(new Object()));
```

The code means that no matter which sentence I delete or doc it, I get the same result (print true)

**Comment by normanmaurer** [comment]
@carryxyh the comment is correct... 

```java
DefaultResourceLeak(
                Object referent,
                ReferenceQueue<Object> refQueue,
                ConcurrentMap<DefaultResourceLeak<?>, LeakEntry> allLeaks) {
            super(referent, refQueue);

            assert referent != null;

            // Store the hash of the tracked object to later assert it in the close(...) method.
            // It's important that we not store a reference to the referent as this would disallow it from
            // be collected via the WeakReference.
            trackedHash = System.identityHashCode(referent);
            allLeaks.put(this, LeakEntry.INSTANCE);
            // Create a new Record so we always have the creation stacktrace included.
            headUpdater.set(this, new Record(Record.BOTTOM));
            this.allLeaks = allLeaks;
        }
```

The referent is the Object that is passed into the constructor here (and not `this`).

**Comment by carryxyh** [comment]
Thank you for your reply.
After I got home last night, I thought about this problem and I understand it. As you said, I understand the reference as 'this', but it should actually be Object.
Thank you.
